src/main/java/com/google/devtools/build/lib/bazel/repository/skylark/SkylarkRepositoryContext.java

```diff
@@ -262,6 +263,29 @@ public void createFileFromTemplate(
     }
   }
 
+  @Override
+  public String readFile(Object path, String encoding, Location location)
+      throws RepositoryFunctionException, EvalException, InterruptedException {
+    SkylarkPath p = getPath("read()", path);
+    Charset charset;
+    if (encoding.equals("")) {
+      charset = StandardCharsets.ISO_8859_1;
+    } else if (encoding.equalsIgnoreCase("utf-8")) {
+      charset = StandardCharsets.UTF_8;
+    } else {
+      throw new EvalException(Location.BUILTIN, "read() only supports encodings in ['utf-8'].");
+    }
+    WorkspaceRuleEvent w =
+        WorkspaceRuleEvent.newReadEvent(
+            p.toString(), encoding, rule.getLabel().toString(), location);
+    env.getListener().post(w);
+    try {
+      return FileSystemUtils.readContent(p.getPath(), charset);
```

Comment:
Starlark strings are raw bytes (https://github.com/bazelbuild/starlark/blob/master/spec.md#strings). Therefore this method should not have any encoding passed in.

Reply:
`repository_ctx.file()` assumes strings are UTF-8:
```java
public void createFile(Object path, String content, Boolean executable, Location location) {
      ...
      try (OutputStream stream = p.getPath().getOutputStream()) {
        stream.write(content.getBytes(StandardCharsets.UTF_8));
      }
```

If the encoding does not default to UTF-8, then `ctx.file("foo", s); ctx.read("foo") == s` will be false.

If we do not allow the user to specify no encoding, then it won't be possible to properly read a binary file.

If you're certain that `ctx.read()` should not have an encoding parameter, should `ctx.file()` also be changed to treat its input as ISO-8859-1 (i.e. plain bytes)?

Reply:
> repository_ctx.file() assumes strings are UTF-8:

That's a bug in `ctx.file()`

> If we do not allow the user to specify no encoding, then it won't be possible to properly read a binary file.

We should read in ISO-8859-1 - then we get raw bytes.

> If you're certain that ctx.read() should not have an encoding parameter, should ctx.file() also be changed to treat its input as ISO-8859-1 (i.e. plain bytes)?

Yes (this is an incompatible change, so the way to do it is to add `legacy_utf8` boolean to `ctx.file`, defaulting to  `true` and go through incompatible change process to make the default `false`.

Reply:
Or rename `ctx.file()` to `ctx.write()` to be symmetrical with `read()` and make these paired functions behave in the new way, and deprecate `file()`

Still requires the incompatible change process to deal with the deprecation, but it's probably more instructive and better named.

Reply:
OK, changes made in new commit:
* Remove `encoding=`, and reads be decoded as ISO-8859-1.
* Add `ctx.write_file()`, which is like `ctx.file()` but encodes with ISO-8859-1.
  * I also had to use a new `workspace_log.proto` message type, because the old `FileEvent` has its contents in `string` instead of `bytes`.
* Renamed new function to `ctx.read_file()` to match.

Reply:
Let's keep the `ctx.file()` and add `legacy_utf8` as I suggested. Why? Migration will be no-op for most (I expect 100% of) people, where with function rename people will have to modify their code.

Let's keep the name `ctx.read()` the function you are adding.

I am not happy about the new event in BEP. Add bytes to `FileEvent` instead.

Reply:
@dslomov Above changes applied.